cdap-app-fabric/src/main/java/co/cask/cdap/route/store/LocalRouteStore.java

```diff
@@ -0,0 +1,127 @@
+ * where performance is not a big issue but we still need persistence of configuration routes across restarts.
+ */
+public class LocalRouteStore implements RouteStore  {
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_INTEGER_TYPE = new TypeToken<Map<String, Integer>>() { }.getType();
+  private static final DatasetId ROUTE_STORE_DATASET_INSTANCE_ID = NamespaceId.SYSTEM.dataset("routestore");
+
+  private final DatasetFramework datasetFramework;
+  private final Transactional transactional;
+
+  @Inject
+  public LocalRouteStore(DatasetFramework datasetFramework, TransactionSystemClient txClient) {
+    this.datasetFramework = datasetFramework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)), RetryStrategies.retryOnConflict(20, 100));
+  }
+
+  private KeyValueTable getRouteTable(DatasetContext context) throws IOException, DatasetManagementException {
+    return DatasetsUtil.getOrCreateDataset(context, datasetFramework, ROUTE_STORE_DATASET_INSTANCE_ID,
+                                           KeyValueTable.class.getName(), DatasetProperties.EMPTY);
+  }
+
+  @Override
+  public void store(final ProgramId serviceId, final RouteConfig routeConfig) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          getRouteTable(context).write(ServiceDiscoverable.getName(serviceId), GSON.toJson(routeConfig.getRoutes()));
+        }
+      });
+    } catch (TransactionFailureException ex) {
+      throw Transactions.propagate(ex);
+    }
+  }
+
+  @Override
+  public void delete(final ProgramId serviceId) throws NotFoundException {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          byte[] key = Bytes.toBytes(ServiceDiscoverable.getName(serviceId));
+          KeyValueTable kvTable = getRouteTable(context);
+          if (kvTable.read(key) == null) {
+            throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+          }
+          kvTable.delete(key);
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e, NotFoundException.class);
+    }
+  }
+
+  @Override
+  public RouteConfig fetch(final ProgramId serviceId) throws NotFoundException {
+    try {
+      return Transactions.execute(transactional, new TxCallable<RouteConfig>() {
+        @Override
+        public RouteConfig call(DatasetContext context) throws Exception {
+          byte[] value = getRouteTable(context).read(ServiceDiscoverable.getName(serviceId));
+          if (value == null || value.length == 0) {
```

Comment:
In the above methods, we get a RouteConfig object which is guaranteed to have a proper map with the values adding up to 100. I can remove this length check in this method, I don't see the value add.